sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/util/S3Util.java

```diff
@@ -0,0 +1,573 @@
+   */
+  public static class S3UtilFactory implements DefaultValueFactory<S3Util> {
+
+    @Override
+    public S3Util create(PipelineOptions options) {
+      S3Options s3Options = options.as(S3Options.class);
+
+      checkArgument(
+          !Strings.isNullOrEmpty(s3Options.getAwsAccessKeyId()), "AWS access key ID is required");
+      checkArgument(
+          !Strings.isNullOrEmpty(s3Options.getAwsSecretAccessKey()),
+          "AWS secret access key is required");
+      checkArgument(!Strings.isNullOrEmpty(s3Options.getAwsRegion()), "AWS region is required");
+
+      return new S3Util(
+          s3Options.getAwsAccessKeyId(),
+          s3Options.getAwsSecretAccessKey(),
+          s3Options.getAwsRegion(),
+          s3Options.getS3StorageClass(),
+          s3Options.getS3UploadBufferSizeBytes(),
+          s3Options.getS3ThreadPoolSize());
+    }
+  }
+
+  // Amazon S3 API docs: Each part must be at least 5 MB in size, except the last part.
+  private static final int MINIMUM_UPLOAD_BUFFER_SIZE_BYTES = 5 * 1024 * 1024;
+  private static final int DEFAULT_UPLOAD_BUFFER_SIZE_BYTES =
+      Runtime.getRuntime().maxMemory() < 512 * 1024 * 1024
+          ? MINIMUM_UPLOAD_BUFFER_SIZE_BYTES
+          : 64 * 1024 * 1024;
+  private static final int MAX_THREADS_PER_CONCURRENT_COPY = 3;
+
+  // Non-final for testing.
+  private AmazonS3 amazonS3;
+  private final String storageClass;
+  private final int s3UploadBufferSizeBytes;
+  private final ExecutorService executorService;
+
+  private S3Util(
+      String awsAccessKeyId,
+      String awsSecretAccessKey,
+      String region,
+      String storageClass,
+      Integer uploadBufferSizeBytes,
+      int threadPoolSize) {
+    checkArgument(!Strings.isNullOrEmpty(awsAccessKeyId), "awsAccessKeyId");
+    checkArgument(!Strings.isNullOrEmpty(awsSecretAccessKey), "awsSecretAccessKey");
+    checkArgument(!Strings.isNullOrEmpty(region));
+
+    AWSStaticCredentialsProvider credentialsProvider =
+        new AWSStaticCredentialsProvider(
+            new BasicAWSCredentials(awsAccessKeyId, awsSecretAccessKey));
+    amazonS3 =
+        AmazonS3ClientBuilder.standard()
+            .withCredentials(credentialsProvider)
+            .withRegion(region)
+            .build();
+
+    this.storageClass = checkNotNull(storageClass, "storageClass");
+
+    if (uploadBufferSizeBytes == null) {
+      uploadBufferSizeBytes = DEFAULT_UPLOAD_BUFFER_SIZE_BYTES;
+    }
+    this.s3UploadBufferSizeBytes =
+        Math.max(MINIMUM_UPLOAD_BUFFER_SIZE_BYTES, uploadBufferSizeBytes);
+
+    checkArgument(threadPoolSize > 0, "threadPoolSize");
+    this.executorService = Executors.newFixedThreadPool(threadPoolSize);
+  }
+
+  @VisibleForTesting
+  public void setAmazonS3Client(AmazonS3 amazonS3) {
+    this.amazonS3 = amazonS3;
+  }
+
+  @VisibleForTesting
+  public int getS3UploadBufferSizeBytes() {
+    return s3UploadBufferSizeBytes;
+  }
+
+  @Override
+  protected void finalize() {
```

Comment:
Hm why not just make this AutoCloseable?

Reply:
This is now folded into S3Filesystem.

Looking at what other FileSystems do:
GcsFileSystem creates a thread pool per batch of work
HadoopFileSystem and LocalFileSystem don't use concurrency

So I'll switch to the GcsFileSystem approach.

Reply:
Done